client-spark/common/src/main/java/org/apache/spark/shuffle/RssStageResubmitManager.java

```diff
@@ -19,51 +19,99 @@
 
-import com.google.common.collect.Sets;
+import org.apache.spark.SparkConf;
+import org.apache.spark.shuffle.stage.RssShuffleStatus;
+import org.apache.spark.shuffle.stage.RssShuffleStatusForReader;
+import org.apache.spark.shuffle.stage.RssShuffleStatusForWriter;
+import org.eclipse.jetty.util.ConcurrentHashSet;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import org.apache.uniffle.common.util.JavaUtils;
-
+/** This class is to manage shuffle status for stage retry */
 public class RssStageResubmitManager {
-
   private static final Logger LOG = LoggerFactory.getLogger(RssStageResubmitManager.class);
-  /** Blacklist of the Shuffle Server when the write fails. */
-  private Set<String> serverIdBlackList;
-  /**
-   * Prevent multiple tasks from reporting FetchFailed, resulting in multiple ShuffleServer
-   * assignments, stageID, Attemptnumber Whether to reassign the combination flag;
-   */
-  private Map<Integer, RssStageInfo> serverAssignedInfos;
 
-  public RssStageResubmitManager() {
-    this.serverIdBlackList = Sets.newConcurrentHashSet();
-    this.serverAssignedInfos = JavaUtils.newConcurrentMap();
+  private final SparkConf sparkConf;
+  private final Map<Integer, RssShuffleStatusForReader> shuffleStatusForReader =
+      new ConcurrentHashMap<>();
+  private final Map<Integer, RssShuffleStatusForWriter> shuffleStatusForWriter =
+      new ConcurrentHashMap<>();
+  private final Map<Integer, Object> shuffleLock = new ConcurrentHashMap<>();
+
+  private final Set<String> blackListedServerIds = new ConcurrentHashSet<>();
+
+  public RssStageResubmitManager(SparkConf sparkConf) {
+    this.sparkConf = sparkConf;
   }
 
-  public Set<String> getServerIdBlackList() {
-    return serverIdBlackList;
+  public Object getOrCreateShuffleLock(int shuffleId) {
+    return shuffleLock.computeIfAbsent(shuffleId, x -> new Object());
   }
 
-  public void resetServerIdBlackList(Set<String> failuresShuffleServerIds) {
-    this.serverIdBlackList = failuresShuffleServerIds;
+  public void clear(int shuffleId) {
+    shuffleStatusForReader.remove(shuffleId);
+    shuffleStatusForWriter.remove(shuffleId);
+    shuffleLock.remove(shuffleId);
   }
 
-  public void recordFailuresShuffleServer(String shuffleServerId) {
-    serverIdBlackList.add(shuffleServerId);
+  public boolean isStageAttemptRetried(int shuffleId, int stageId, int stageAttemptNumber) {
+    RssShuffleStatus readerShuffleStatus = shuffleStatusForReader.get(shuffleId);
+    RssShuffleStatus writerShuffleStatus = shuffleStatusForWriter.get(shuffleId);
+    if (readerShuffleStatus == null && writerShuffleStatus == null) {
+      return false;
+    }
+    if (readerShuffleStatus != null
+        && readerShuffleStatus.isStageAttemptRetried(stageAttemptNumber)) {
+      return true;
+    }
+    if (writerShuffleStatus != null
+        && writerShuffleStatus.isStageAttemptRetried(stageAttemptNumber)) {
+      return true;
+    }
+    return false;
   }
```

Comment:
In one case, the Reader triggers retry, and the retry is recorded. After the Writer fails to write data for several times, the retry is triggered. However, this method returns that the retry has been performed.

Reply:
Yes. The same stageIdAttemptNumber retry will ocurr one time, is this incorrect? @yl09099